contrib/config/consul/consul.go

```diff
@@ -0,0 +1,166 @@
+	if err != nil {
+		return nil, gerror.Wrapf(err, `create consul client failed with config: %+v`, config.Params)
+	}
+
+	if err = client.addWatcher(); err != nil {
+		return nil, gerror.Wrapf(err, `consul client add watcher failed with config: %+v`, config.Params)
+	}
+
+	return client, nil
+}
+
+// Available checks and returns the backend configuration service is available.
+// The optional parameter `resource` specifies certain configuration resource.
+//
+// Note that this function does not return error as it just does simply check for
+// backend configuration service.
+func (c *Client) Available(ctx context.Context, resource ...string) (ok bool) {
+	if len(resource) == 0 && !c.value.IsNil() {
+		return true
+	}
+
+	_, _, err := c.client.KV().Get(c.config.Path, nil)
+
+	return err == nil
+}
+
+// Get retrieves and returns value by specified `pattern` in current resource.
+// Pattern like:
+// "x.y.z" for map item.
+// "x.0.y" for slice item.
+func (c *Client) Get(ctx context.Context, pattern string) (value interface{}, err error) {
+	if c.value.IsNil() {
+		if err = c.updateLocalValue(); err != nil {
+			return nil, err
+		}
+	}
+	return c.value.Val().(*gjson.Json).Get(pattern).Val(), nil
+}
+
+// Data retrieves and returns all configuration data in current resource as map.
+// Note that this function may lead lots of memory usage if configuration data is too large,
+// you can implement this function if necessary.
+func (c *Client) Data(ctx context.Context) (data map[string]interface{}, err error) {
+	if c.value.IsNil() {
+		if err = c.updateLocalValue(); err != nil {
+			return nil, err
+		}
+	}
+	return c.value.Val().(*gjson.Json).Map(), nil
+}
+
+func (c *Client) updateLocalValue() (err error) {
+	content, _, err := c.client.KV().Get(c.config.Path, nil)
+	if err != nil {
+		return gerror.Wrapf(err, `get config from consul path [%+v] failed`, c.config.Path)
+	}
+	if content == nil {
+		return fmt.Errorf(`get config from consul path [%+v] value is nil`, c.config.Path)
+	}
+	return c.doUpdate(content.Value)
+}
+
+func (c *Client) doUpdate(content []byte) (err error) {
+	var j *gjson.Json
+	if j, err = gjson.LoadContent(content); err != nil {
+		return gerror.Wrapf(err,
+			`parse config map item from consul path [%+v] failed`, c.config.Path)
+	}
+	c.value.Set(j)
+	return nil
+}
+
+func (c *Client) addWatcher() (err error) {
+	if !c.config.Watch {
+		return nil
+	}
+
+	plan, err := watch.Parse(map[string]interface{}{
+		"type": "key",
+		"key":  c.config.Path,
+	})
+	if err != nil {
+		return gerror.Wrapf(err, `watch config from consul path %+v failed`, c.config.Path)
+	}
+
+	plan.Handler = func(idx uint64, raw interface{}) {
+		var v *api.KVPair
+		if raw == nil {
+			// nil is a valid return value
+			v = nil
+			return
+		}
+		var ok bool
+		if v, ok = raw.(*api.KVPair); !ok {
+			return
+		}
+
+		if err := c.doUpdate(v.Value); err != nil {
+			g.Log(LoggerName).Errorf(context.Background(),
```

Comment:
Any logging operation please use an interface, and passed implements from user, not using `g.Log` directly.

Reply:
OK